contracts/staking/utils/StakingRewardsPerEpoch.sol

```diff
@@ -0,0 +1,166 @@
+
+	modifier updateReward(address account) {
+		_updateReward(account);
+		_;
+	}
+
+	/**
+	 * @dev A classic ERC20 method that allows staker balance acquiring.
+	 * @param account A staker address
+	 */
+	function balanceOf(address account) external view returns (uint256) {
+		return _balanceOf(account);
+	}
+
+	function _balanceOf(address account) internal view returns (uint256) {
+		// The resulting balance of any user is the sum of an active earning balance
+		// and pending waiting balance.
+		return stakersInfo[account].balance + stakersInfo[account].pendingStake;
+	}
+
+	function _getRewardPerTokenPerUser(address _account)
+		internal
+		view
+		returns (uint256)
+	{
+		// The userEpochIndex is used to calculate a reward per toker per epoch
+		// getting into account passed epochs.
+		uint256 userEpochIndex = stakersInfo[_account].indexOfLastEpochStaked + 1;
+		if (lastEpochIndex > userEpochIndex) {
+			// Here we calculate the reward getting into account passed epochs.
+			return rewardsPerTokenAt[lastEpochIndex] - rewardsPerTokenAt[userEpochIndex];
+		} else {
+			return 0;
+		}
+	}
+
+	/**
+	 * @dev The function allows anyone to calculate the exact amount of reward
+	 * earned per epochs passed.
+	 * @param account A staker address
+	 */
+	function earned(address account) public view returns (uint256) {
+		return
+			(stakersInfo[account].balance * _getRewardPerTokenPerUser(account))
+				/ PRECISION
+				+ stakersInfo[account].reward;
+	}
+
+	function _addPendingStakesToBalanceOnEpoch(address _account) internal {
+		// If stakers balance weren't updated when he staked and the staked
+		// amount is greater than 0, then update active earning balance, nullify
+		// pending one and update global sum of pending stakes.
+		if (stakersInfo[_account].indexOfLastEpochStaked != lastEpochIndex
+				&& stakersInfo[_account].pendingStake > 0) {
+			stakersInfo[_account].balance += stakersInfo[_account].pendingStake;
+			stakersInfo[_account].pendingStake = 0;
+			pendingStakes -= stakersInfo[_account].pendingStake;
+		}
+	}
+
+	// this function updates the reward for the specific user
+	function _updateReward(address _account) internal virtual {
+		_addPendingStakesToBalanceOnEpoch(_account);
+		stakersInfo[_account].reward = earned(_account);
+	}
+
+	// This function adds the sum given in reward parameter to the distribution
+	// queue.
+	function _notifyRewardAmount(uint256 reward) internal {
+		// update cumulative rewards
+		rewardsPerTokenAt.push(
+			rewardsPerTokenAt[rewardsPerTokenAt.length - 1]
+				+ (reward * PRECISION) / totalSupply
+		);
+		// turn pending stakes to active stakes
+		totalSupply += pendingStakes;
+		pendingStakes = 0;
+
+		// update epoch count
+		lastEpochIndex++;
+	}
+
+	function _withdraw(address _from, uint256 _amount) internal virtual {
+		// if there are any pending stake for _from
+		if (stakersInfo[_from].pendingStake > 0) {
+			// if requested sum to withdraw is higher than actual pending stake balance
```

Comment:
Opposite. If pending stakes are higher or equal to requested withdrawal amount -> subtract the whole amount, otherwise subtract just whatever pending stake there is